src/torchmetrics/detection/mean_ap.py

```diff
@@ -604,23 +600,23 @@ def _evaluate_image(
         # load computed ious
         ious = ious[idx, class_id][:, gtind] if len(ious[idx, class_id]) > 0 else ious[idx, class_id]
 
-        nb_iou_thrs = len(self.iou_thresholds)
         nb_gt = len(gt)
         nb_det = len(det)
         gt_matches = torch.zeros((nb_iou_thrs, nb_gt), dtype=torch.bool, device=self.device)
-        det_matches = torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool, device=self.device)
         gt_ignore = ignore_area_sorted
-        det_ignore = torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool, device=self.device)
+
+        iou_thresholds = torch.tensor(self.iou_thresholds, device=self.device)
 
         if torch.numel(ious) > 0:
-            for idx_iou, t in enumerate(self.iou_thresholds):
-                for idx_det, _ in enumerate(det):
-                    m = MeanAveragePrecision._find_best_gt_match(t, gt_matches, idx_iou, gt_ignore, ious, idx_det)
-                    if m == -1:
-                        continue
-                    det_ignore[idx_iou, idx_det] = gt_ignore[m]
-                    det_matches[idx_iou, idx_det] = 1
-                    gt_matches[idx_iou, m] = 1
+            best_matches = self._find_best_gt_matches(iou_thresholds, gt_matches, gt_ignore, ious)
+            _zero_tensor = torch.tensor(0, dtype=torch.bool, device=self.device)
+            _one_tensor = torch.tensor(1, dtype=torch.bool, device=self.device)
+            det_ignore = torch.where(
+                best_matches != -1, gt_ignore[best_matches.clamp(max=gt_ignore.shape[0] - 1)], _zero_tensor
+            )
```

Comment:
this does not work in torch 1.7 (apparently) because the 
```RuntimeError: "where_cpu" not implemented for 'Bool'```
the simple solution is casting to float before and bool afterwards, but not sure if we should just skip 1.7
```suggestion
            det_ignore = torch.where(
                best_matches != -1, gt_ignore[best_matches.clamp(max=gt_ignore.shape[0] - 1)].float(), _zero_tensor.float()
            ).bool()
```

Reply:
Oh, good catch. Maybe I can encapsulate this into a method and within that method differentiate between `torch==1.7` and other versions? Not sure how much this affects the performance (likely not much).

Reply:
At this point I think we could also drop 1.7 compatibility. It is already quite old...

Reply:
If we settle for 1.8 being minimum we would still be compatible with 5 releases of pytorch.

Reply:
Yes, I think aiming for 5 major releases would be a good compromise (in PL we do fewer). This would also make it easier to implement new features since at some point you are quite limited with the older versions of torch

Reply:
yes, I am fine with set min version 1.8, just lets do it in separate PR :)

Reply:
just created issue for future reference: https://github.com/Lightning-AI/metrics/issues/1262

Reply:
and pr: https://github.com/Lightning-AI/metrics/pull/1263